## YOLOX evaluation on COCO dataset
To evaluate YOLOX on COCO dataset , First using evaluation set of COCO dataset and then using test set of COCO dataset.

## Evaluation set using my edited yolox fork repo
Clone the repo

In [1]:
!git clone https://github.com/KerolosAtef/YOLOX-fork.git

Cloning into 'YOLOX-fork'...
remote: Enumerating objects: 385, done.
remote: Counting objects: 100% (385/385), done.
remote: Compressing objects: 100% (231/231), done.
remote: Total 385 (delta 121), reused 371 (delta 109), pack-reused 0
Receiving objects: 100% (385/385), 5.14 MiB | 18.98 MiB/s, done.
Resolving deltas: 100% (121/121), done.


## Downloading the weights and the dataset

In [2]:
!wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_s.pth
!wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_m.pth
!wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_l.pth
!wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_x.pth
!wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_darknet.pth

--2023-05-30 03:57:07--  https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_s.pth
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/386811486/42c4cb47-f94e-475b-a3a2-57f31f26fa5d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230530%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230530T035707Z&X-Amz-Expires=300&X-Amz-Signature=42a06611c8cbf973520f6eb89d8e8c5b7792edbf682ad66d40ce51dbe830f44a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=386811486&response-content-disposition=attachment%3B%20filename%3Dyolox_s.pth&response-content-type=application%2Foctet-stream [following]
--2023-05-30 03:57:07--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/386811486/42c4cb47-f94e-475b-a3a2-57f31f26fa5d?X-Am

### Downloading the test dataset

In [12]:
!wget http://images.cocodataset.org/zips/test2017.zip

--2023-05-30 04:30:10--  http://images.cocodataset.org/zips/test2017.zip
Resolving images.cocodataset.org (images.cocodataset.org)... 52.217.102.28, 52.217.162.65, 52.217.69.212, ...
Connecting to images.cocodataset.org (images.cocodataset.org)|52.217.102.28|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6646970404 (6.2G) [application/zip]
Saving to: ‘test2017.zip’

test2017.zip        100%[===================>]   6.19G  56.1MB/s    in 1m 44s  

2023-05-30 04:31:54 (60.8 MB/s) - ‘test2017.zip’ saved [6646970404/6646970404]



### Downloading the validation dataset

In [4]:
! wget http://images.cocodataset.org/zips/val2017.zip

--2023-05-30 03:58:35--  http://images.cocodataset.org/zips/val2017.zip
Resolving images.cocodataset.org (images.cocodataset.org)... 54.231.171.129, 52.217.118.217, 52.217.166.113, ...
Connecting to images.cocodataset.org (images.cocodataset.org)|54.231.171.129|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 815585330 (778M) [application/zip]
Saving to: ‘val2017.zip’

val2017.zip         100%[===================>] 777.80M  94.0MB/s    in 8.4s    

2023-05-30 03:58:43 (92.7 MB/s) - ‘val2017.zip’ saved [815585330/815585330]



### Installing the requirements

In [5]:
import os 
os.chdir('/kaggle/working/YOLOX-fork')
!pip3 install -v -e .

Using pip 23.1.2 from /opt/conda/lib/python3.10/site-packages/pip (python 3.10)
Obtaining file:///kaggle/working/YOLOX-fork
  Running command python setup.py egg_info
  running egg_info
  creating /tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info
  writing /tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/PKG-INFO
  writing dependency_links to /tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/dependency_links.txt
  writing requirements to /tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/requires.txt
  writing top-level names to /tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/top_level.txt
  writing manifest file '/tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  writing manifest file '/tmp/pip-pip-egg-info-8biqf4ie/yolox.egg-info/SOURCES.txt'
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 2.8 MB/s eta 0:00:00
  Running command pip subprocess to install build dependen

### Unzipping the datasets and moving them to the correct directory

In [13]:
!unzip -q /kaggle/working/val2017.zip -d /kaggle/working/YOLOX-fork/datasets/COCO
!unzip -q /kaggle/working/test2017.zip -d /kaggle/working/YOLOX-fork/datasets/COCO

[/kaggle/working/test2017.zip]
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of /kaggle/working/test2017.zip or
        /kaggle/working/test2017.zip.zip, and cannot find /kaggle/working/test2017.zip.ZIP, period.


### Downloading the annotations for the validation dataset and moving them to the correct directory

In [8]:
! wget http://images.cocodataset.org/annotations/annotations_trainval2017.zip

--2023-05-29 23:58:39--  http://images.cocodataset.org/annotations/annotations_trainval2017.zip
Resolving images.cocodataset.org (images.cocodataset.org)... 3.5.20.197, 52.217.93.204, 52.217.97.124, ...
Connecting to images.cocodataset.org (images.cocodataset.org)|3.5.20.197|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 252907541 (241M) [application/zip]
Saving to: ‘annotations_trainval2017.zip’

annotations_trainva 100%[===================>] 241.19M  45.7MB/s    in 10s     

2023-05-29 23:58:50 (24.0 MB/s) - ‘annotations_trainval2017.zip’ saved [252907541/252907541]



In [9]:
!unzip -q /kaggle/working/YOLOX-fork/annotations_trainval2017.zip -d /kaggle/working/YOLOX-fork/datasets/COCO

### Evaluating the model on the validation dataset using yolox-eval script

## YOLOX-s

In [5]:
from torchsummary import summary
from yolox.exp import get_exp
from yolox.utils import fuse_model, get_model_info, postprocess, vis
exp = get_exp(exp_name="yolox-s")
model = exp.get_model()
"Model Summary: {}".format(get_model_info(model, exp.test_size))

'Model Summary: Params: 8.97M, Gflops: 26.93'

In [10]:
! python -m yolox.tools.eval -n  yolox-s -c /kaggle/working/yolox_s.pth -b 64 -d 0 --conf 0.001 [--fp16] [--fuse]

/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tenso

## Highlighting the important results from the evaluation
### Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.405
| Model                                    |size |mAP<sup>val<br>0.5:0.95 |   Params<br>(M)   | FLOPs<br>(G) | weights |
|------------------------------------------|:---: | :---:    |:-----------------:|:------------:| :----: |
| [YOLOX-s](./exps/default/yolox_s.py)     |640  |40.5 |         9.0       |     26.8     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_s.pth) |
| [YOLOX-s-exp](./exps/default/yolox_s.py) |640  |40.5 |   8.97      |    26.93     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_s.pth) |


## YOLOX-m

In [6]:
from torchsummary import summary
from yolox.exp import get_exp
from yolox.utils import fuse_model, get_model_info, postprocess, vis
exp = get_exp(exp_name="yolox-m")
model = exp.get_model()
"Model Summary: {}".format(get_model_info(model, exp.test_size))

'Model Summary: Params: 25.33M, Gflops: 73.98'

In [12]:
! python -m yolox.tools.eval -n  yolox-m -c /kaggle/working/yolox_m.pth -b 64 -d 0 --conf 0.001 [--fp16] [--fuse]

/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tenso

## Highlighting the important results from the evaluation
### Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.469
| Model                                    |size | mAP<sup>val<br>0.5:0.95 | Params<br>(M) | FLOPs<br>(G) |                                            weights                                             |
|------------------------------------------|:---: |:-----------------------:|:-------------:|:------------:|:----------------------------------------------------------------------------------------------:|
| [YOLOX-m](./exps/default/yolox_m.py)     |640  |          46.9           |     25.3      |     73.8     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_m.pth) |
| [YOLOX-m-exp](./exps/default/yolox_m.py) |640  |          46.9           |     25.33     |    73.98     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_m.pth) |

## YOLOX-l

In [8]:
from torchsummary import summary
from yolox.exp import get_exp
from yolox.utils import fuse_model, get_model_info, postprocess, vis
exp = get_exp(exp_name="yolox-l")
model = exp.get_model()
"Model Summary: {}".format(get_model_info(model, exp.test_size))

'Model Summary: Params: 54.21M, Gflops: 156.01'

In [13]:
! python -m yolox.tools.eval -n  yolox-l -c /kaggle/working/yolox_l.pth -b 64 -d 0 --conf 0.001 [--fp16] [--fuse]

/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tenso

## Highlighting the important results from the evaluation
### Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.497
| Model                                    |size | mAP<sup>val<br>0.5:0.95 | Params<br>(M) | FLOPs<br>(G) |                                            weights                                             |
|------------------------------------------|:---: |:-----------------------:|:-------------:|:------------:|:----------------------------------------------------------------------------------------------:|
| [YOLOX-l](./exps/default/yolox_l.py)     |640  |          49.7           |     54.2      |    155.6     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_l.pth) |
| [YOLOX-l-exp](./exps/default/yolox_l.py) |640  |          49.7           |     54.21     |    156.01    | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_l.pth) |

## YOLOX-x

In [9]:
from torchsummary import summary
from yolox.exp import get_exp
from yolox.utils import fuse_model, get_model_info, postprocess, vis
exp = get_exp(exp_name="yolox-x")
model = exp.get_model()
"Model Summary: {}".format(get_model_info(model, exp.test_size))

'Model Summary: Params: 99.07M, Gflops: 282.46'

In [14]:
! python -m yolox.tools.eval -n  yolox-x -c /kaggle/working/yolox_x.pth -b 64 -d 0 --conf 0.001 [--fp16] [--fuse]

/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tenso

## Highlighting the important results from the evaluation
### Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.497
| Model                                    |size | mAP<sup>val<br>0.5:0.95 | Params<br>(M) | FLOPs<br>(G) |                                            weights                                             |
|------------------------------------------|:---: |:-----------------------:|:-------------:|:------------:|:----------------------------------------------------------------------------------------------:|
| [YOLOX-x](./exps/default/yolox_x.py)     |640  |          51.1           |     99.1      |    281.9     | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_x.pth) |
| [YOLOX-x-exp](./exps/default/yolox_x.py) |640  |          51.1           |     99.07     |    282.46    | [github](https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_x.pth) |

## We can see that all the results is True with some small differences in the number of parameters and the FLOPs which is due to the different hardware used for the evaluation they used Nvidia Tesla V100 and I used Nvidia P100 from Kaggle.

## Evaluation on the COCO test set
To evaluate the model against the tests set we need to download the test set folders and the annotations file which is not available to the public.<br>
The only way to get the test set is to register for the COCO challenge and download the test set from the challenge website and prepare the output annotations file and use Coco dataset API to get the scores from there website<br>
**According to that I will not be able to evaluate the model against the test set,but I will extract the results to a json file but not submitted it because it will be a fake submission.**

## First install the sahi to prepare the json file in the coco format

In [15]:
!pip install sahi

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 kB 4.4 MB/s eta 0:00:00


## I will add a new function to the demo.py file to extract the results to a json file in the coco format.
## This is a detailed steps, but the full code will be added to the demo.py file on my forked repo.

In [ ]:
# this part will be added to the demo.py file at the beginning of the file after the imports
COCO_dataset_classes={0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
from sahi.utils.coco import Coco, CocoCategory, CocoImage, CocoAnnotation
from sahi.utils.file import save_json
coco = Coco()
save_path = 'datasets/annotations/custom.json'
for i in range(80):
    coco.add_category(CocoCategory(id=i, name=COCO_dataset_classes[i]))
## Then this function will be added to the Predictor class in the demo.py file
def add_image(self, img_info, output):
    # id_str =img_info['file_name'][:-4]
    # id=''
    # for i in id_str:
    #     if i.isdigit() and i == '0':
    #         id+=i
    coco_image = CocoImage(file_name=img_info['file_name'], height=img_info['height'], width=img_info['width'],
                           id=img_info['id'])
    ratio = img_info["ratio"]
    img = img_info["raw_img"]
    if output is None:
        return img
    output = output.cpu()

    bboxes = output[:, 0:4]

    # preprocessing: resize
    bboxes /= ratio
    cls = output[:, 6]
    for i in range(len(bboxes)):
        coco_image.add_annotation(
            CocoAnnotation(
                bbox=bboxes[i].numpy(),
                category_id=cls[i].item(),
                category_name=self.cls_names[int(cls[0].item())],
            )
        )
    coco.add_image(coco_image)

# I will call add_image function inside image_demo function with this line
# predictor.add_image(img_info,outputs[0])
# I will comment the part of visualization of the results because it will not be needed and consume time and memory
# Then I will save the json file at the end of the main function with this line

## The COCO dataset has some parts in the annotation file such as info,licenses so I will copy them from the train annotation file to the test annotation file

For the limited time and resources I will evaluate the yolox-s model only

In [ ]:
!python tools/demo.py image -n yolox-s -c /kaggle/working/yolox_s.pth --path datasets/COCO/test2017 --conf 0.25 --nms 0.45 --tsize 640 --device gpu

## Link to the json file that contains the results of yolox-s model on the test set in the coco format [Download](https://drive.google.com/file/d/1DBqsaUCjC9aeznj37Pv_MJ-HvNDSVaMC/view?usp=share_link)